core/src/main/java/org/elasticsearch/discovery/zen/ZenDiscovery.java

```diff
@@ -682,171 +670,90 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
 
 
-                    // some optimizations to make sure we keep old objects where possible
-                    ClusterState.Builder builder = ClusterState.builder(updatedState);
+                // some optimizations to make sure we keep old objects where possible
+                ClusterState.Builder builder = ClusterState.builder(newClusterState);
 
-                    // if the routing table did not change, use the original one
-                    if (updatedState.routingTable().version() == currentState.routingTable().version()) {
-                        builder.routingTable(currentState.routingTable());
-                    }
-                    // same for metadata
-                    if (updatedState.metaData().version() == currentState.metaData().version()) {
-                        builder.metaData(currentState.metaData());
-                    } else {
-                        // if its not the same version, only copy over new indices or ones that changed the version
-                        MetaData.Builder metaDataBuilder = MetaData.builder(updatedState.metaData()).removeAllIndices();
-                        for (IndexMetaData indexMetaData : updatedState.metaData()) {
-                            IndexMetaData currentIndexMetaData = currentState.metaData().index(indexMetaData.index());
-                            if (currentIndexMetaData != null && currentIndexMetaData.isSameUUID(indexMetaData.indexUUID()) &&
-                                    currentIndexMetaData.version() == indexMetaData.version()) {
-                                // safe to reuse
-                                metaDataBuilder.put(currentIndexMetaData, false);
-                            } else {
-                                metaDataBuilder.put(indexMetaData, false);
-                            }
+                // if the routing table did not change, use the original one
+                if (newClusterState.routingTable().version() == currentState.routingTable().version()) {
+                    builder.routingTable(currentState.routingTable());
+                }
+                // same for metadata
+                if (newClusterState.metaData().version() == currentState.metaData().version()) {
+                    builder.metaData(currentState.metaData());
+                } else {
+                    // if its not the same version, only copy over new indices or ones that changed the version
+                    MetaData.Builder metaDataBuilder = MetaData.builder(newClusterState.metaData()).removeAllIndices();
+                    for (IndexMetaData indexMetaData : newClusterState.metaData()) {
+                        IndexMetaData currentIndexMetaData = currentState.metaData().index(indexMetaData.index());
+                        if (currentIndexMetaData != null && currentIndexMetaData.isSameUUID(indexMetaData.indexUUID()) &&
+                                currentIndexMetaData.version() == indexMetaData.version()) {
```

Comment:
I think this check is long enough to warrant a separate (and unit testable!) method for checking if the index metadata is the same